paddle/fluid/framework/ir/graph_pattern_detector.cc

```diff
@@ -869,6 +872,33 @@ PDNode *patterns::FC::operator()(paddle::framework::ir::PDNode *x,
   }
 }
 
+PDNode *patterns::FCMKLDNN::operator()(paddle::framework::ir::PDNode *x,
```

Comment:
Since we have `PDNode *patterns::FC::operator()`, why we need `PDNode *patterns::FCMKLDNN::operator` again?
We don't have `PDNode *patterns::xxxMKLDNN:: operator()` in this file.

Reply:
It's because FC pattern detector searches for mul + elementwise_add pattern, while FCMKLDNN searches for an FC op pattern.

Reply:
1. Yes, it is different:
- Paddle's FC Pattern searches for Mul Op followed by Elementwise Add Op.
- MKL-DNN's FC Pattern searches just for FC Op that was created by fc_fuse_pass. (Not for mul/elementwise_add)
2. It is not enough to set use_mkldnn to true, because the weights ("W") of the operator need to be transposed.

Reply:
The reason why you transpose the weights of this Op in the pass is to avoid duplicated transpose in the Op's `compute()`. How much elapsed time of this weight transpose?

However, from framework perspective, `xxx_mkldnn_op` should have the same behavior with `xxx_op`, they could have different kernels but with the same input/output/weights, i.e, fc's weights should be transposed in the mkldnn kernel. 

Currently, 
- this fc_mkldnn_op is hard to extend to training. If fc_mkldnn_op wants to support training, how do you implement with it?
- `It is not enough to set use_mkldnn to true` may make users confused.

@jianhang-liu How do you think about it?

Reply:
If I may add anything:
The in-kernel transpose incurs so much overhead that even a kernel operating on non-transposed weights would execute faster. This is the case due to duplicated memory reads and writes when the weights data needs first to be read and written (invalidating the cache) for the purpose of transpose, and then read again for the purpose of fc computations.

What I meant by `It is not enough to set use_mkldnn to true` is that from programmer's perspective it is also necessary to transpose the kernel, but from user's perspective it's much easier.

When the `use_mkldnn` is set to true, the pass transposes the weights and **only then** enables the fc operator.

Reply:
Do you need another weights transpose before calling`SaveOptimModel`? Otherwise if we reload the `saved_optimized_model`, it will crashes.

Reply:
There is no need to do that, because if we save the optimized model with all the passes applied, once it will be loaded again it will execute just fine in MKL-DNN environment, because the passes would be already applied and the weights transposed. 

There is no point in running `saved optimized model` in other environment anyway, because the passes such as conv + batch_norm + bias have already introduced changes which are only applicable in MKL-DNN-only environment (there is no support for bias in reference conv).

Is that what you meant?

Reply:
Got it for Fp32, but for INT8,  we will running `saved optimized model` in other environment. #17097 base on this PR and use weights transpose as well.

Reply:
As far as I understand, if kernel from #17097 will be accustomed to use transposed weights, then everything should be set and ready for running `saved optimized model` in int8 python environment. If not, it is always possible to transpose the weights back using transpiler.

Reply:
Current implementation doesn't modify weights. It transposes them and stores them internally in execution context.

Reply:
> Since we have `PDNode *patterns::FC::operator()`, why we need `PDNode *patterns::FCMKLDNN::operator` again?
> We don't have `PDNode *patterns::xxxMKLDNN:: operator()` in this file
The pattern is needed to detect Fully-Connected layer created by `fc_fuse_pass`.